packages/headless/src/features/folding/folding-slice.ts

```diff
@@ -107,34 +185,90 @@ export const foldingReducer = createReducer(
         state.collections = state.enabled
           ? createCollections(
               payload.response.results as ResultWithFolding[],
-              state.fields
+              state.fields,
+              state.filterFieldRange - collectionMoreResultsAvailableBuffer
             )
-          : [];
+          : {};
       })
       .addCase(fetchMoreResults.fulfilled, (state, {payload}) => {
         state.collections = state.enabled
-          ? [
+          ? {
               ...state.collections,
               ...createCollections(
                 payload.response.results as ResultWithFolding[],
-                state.fields
+                state.fields,
+                state.filterFieldRange - collectionMoreResultsAvailableBuffer
               ),
-            ]
-          : [];
+            }
+          : {};
       })
       .addCase(registerFolding, (state, {payload}) =>
         state.enabled
           ? state
           : {
               enabled: true,
-              collections: [],
+              collections: {},
               fields: {
                 collection: payload.collectionField ?? state.fields.collection,
                 parent: payload.parentField ?? state.fields.parent,
                 child: payload.childField ?? state.fields.child,
               },
-              numberOfFoldedResults:
-                payload.numberOfFoldedResults ?? state.numberOfFoldedResults,
+              filterFieldRange: payload.numberOfFoldedResults
+                ? payload.numberOfFoldedResults +
+                  collectionMoreResultsAvailableBuffer
+                : state.filterFieldRange,
             }
       )
+      .addCase(loadCollection.pending, (state, {meta}) => {
+        const collectionId = meta.arg;
+
+        return {
+          ...state,
```

Comment:
You don't need to return the whole state like a typical reducer, since toolkit uses immer https://redux-toolkit.js.org/api/createReducer#direct-state-mutation

That should work, no return necessary
```
const collectionId = meta.arg;
state.collections = {...}
```

Probably even this would work, to be tested:
```
const collectionId = meta.arg;
state.collections[collectionId].isLoadingMoreResults = true;
```

same for the similar cases